app/lib/actions/katello/host/update_system_purpose.rb

```diff
@@ -0,0 +1,28 @@
+module Actions
+  module Katello
+    module Host
+      class UpdateSystemPurpose < Actions::EntryAction
+        def plan(host, service_level, purpose_role, purpose_usage)
+          if host.content_facet
+            host.subscription_facet.service_level = service_level
+            host.subscription_facet.purpose_role = purpose_role
+            host.subscription_facet.purpose_usage = purpose_usage
+
+            host.update_candlepin_associations
```

Comment:
subscription facet has a before_update callback which calls update_candlepin_associations, so you shouldn't need to call it here. See subscription_facet_host_extensions.rb for the definition

You can also update multiple attributes at once: `host.subscription_facet.update(service_level: 'foo', purpose_role: 'bar')` etc

Reply:
After our talk on IRC kept it the way it was since we have to update purpose addons in a different way and host.save works better